### PR TITLE
fix: Do not enforce all shards tracking in stateless validation

### PR DESCRIPTION
**Pull request by staffik**
Follow-up to: https://github.com/near/nearcore/pull/11766

Tested in testnet.
* I've built once with `statelessnet_protocol` feature  (so that binary protocol version is 82) and confirmed that the protocol version we used in the check comes from the chain (it was 69).
* In another test I subtracted 1 from the protocol version (so that it becomes 68) and confirmed that the check kicks in.
* It panics within a second from starting neard. I think it is good enough. All logs:
```
$ time ./target/release/neard --home ~/.near run
2024-07-12T15:44:19.445537Z  INFO neard: version="trunk" build="1.36.1-905-g08f4e9781-modified" latest_protocol=69
2024-07-12T15:44:19.445781Z  INFO config: Validating Config, extracted from config.json...
2024-07-12T15:44:19.446105Z  WARN genesis: Skipped genesis validation
2024-07-12T15:44:19.446123Z  INFO config: Validating Genesis config and records. This could take a few minutes...
2024-07-12T15:44:19.447017Z  INFO config: All validations have passed!
2024-07-12T15:44:19.447080Z  WARN neard: Running a neard executable which wasn’t built with `make release` command isn’t supported on testnet.
2024-07-12T15:44:19.447089Z  WARN neard: Note that `cargo build --release` builds lack optimisations which may be needed to run properly on testnet
2024-07-12T15:44:19.447096Z  WARN neard: Consider recompiling the binary using `make release` command.
2024-07-12T15:44:19.449330Z  INFO neard: Reset the config "/home/ubuntu/.near/log_config.json" because the config file doesn't exist. err=Os { code: 2, kind: NotFound, message: "No such file or directory" }
2024-07-12T15:44:19.449461Z  INFO config: Validating Config, extracted from config.json...
2024-07-12T15:44:19.449523Z  INFO neard: Hot loading validator key /home/ubuntu/.near/validator_key.json.
2024-07-12T15:44:19.449567Z  INFO near_o11y::reload: Updated the logging layer according to `log_config.json`
2024-07-12T15:44:19.449598Z  INFO db_opener: Opening NodeStorage path="/home/ubuntu/.near/data" cold_path="none"
2024-07-12T15:44:19.449626Z  INFO db: Opened a new RocksDB instance. num_instances=1
2024-07-12T15:44:19.481493Z  INFO db: Closed a RocksDB instance. num_instances=0
2024-07-12T15:44:19.481514Z  INFO db_opener: The database exists. path=/home/ubuntu/.near/data
2024-07-12T15:44:19.481528Z  INFO db: Opened a new RocksDB instance. num_instances=1
2024-07-12T15:44:19.769204Z  INFO db: Closed a RocksDB instance. num_instances=0
2024-07-12T15:44:19.769252Z  INFO db: Opened a new RocksDB instance. num_instances=1
2024-07-12T15:44:19.798895Z  INFO db: Closed a RocksDB instance. num_instances=0
2024-07-12T15:44:19.798928Z  INFO db: Opened a new RocksDB instance. num_instances=1
thread 'actix-rt|system:0|arbiter:5' panicked at chain/client/src/client_actor.rs:333:17:
The `chain_id` field specified in genesis is among mainnet/betanet/testnet, so validator must track all shards. Please change `tracked_shards` field in config.json to be any non-empty vector
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: actix_rt::arbiter::ArbiterHandle::spawn_fn::{{closure}}
   3: tokio::runtime::task::raw::poll
   4: tokio::task::local::LocalSet::tick
   5: tokio::task::local::LocalSet::run_until::{{closure}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Aborted

real    0m0.696s
```

There is one caveat. The check was originally in `load_config` function and there exist a command:
```
impl ValidateConfigCommand {
    pub(super) fn run(&self, home_dir: &Path) -> anyhow::Result<()> {
        nearcore::config::load_config(home_dir, GenesisValidationMode::Full)?;
        Ok(())
    }
}
```
We can duplicate the code a bit and have the check in both places. Wdyt?

### CODE REVIEW

File: chain/client/src/client_actor.rs [hunk 1]
```diff
@@ -317,6 +317,21 @@ impl ClientActorInner {
...
+                .into_chain_error()?;
+            // Before stateless validation we require validators to track all shards, see
+            // https://github.com/near/nearcore/issues/7388
+            if !ProtocolFeature::StatelessValidationV0.enabled(protocol_version)
+                && config.tracked_shards.is_empty()
+                && matches!(
+                    config.chain_id.as_ref(),
+                    near_primitives::chains::MAINNET | near_primitives::chains::TESTNET
+                ) {
+                panic!("The `chain_id` field specified in genesis is among mainnet/betanet/testnet, so validator must track all shards. Please change `tracked_shards` field in config.json to be any non-empty vector");
```
Comment (wacban) [inline chain/client/src/client_actor.rs:333]: Why mention betanet? I don't think it exists anymore.

File: chain/client/src/client_actor.rs [hunk 2]
```diff
@@ -317,6 +317,21 @@ impl ClientActorInner {
     ) -> Result<Self, Error> {
         if let Some(vs) = &client.validator_signer.get() {
             info!(target: "client", "Starting validator node: {}", vs.validator_id());
+            let head = client.chain.head()?;
```
Comment (wacban) [inline chain/client/src/client_actor.rs:320]: nit: Maybe move all of this logic to a separate method?

**Review by wacban — APPROVED** [review]
LGTM

What about validator key hot swap? Should this check also be triggered then? 

We also discussed forcing single shard tracking for stateless validators - are you going to do that separately? 

Should this be cherry picked into the 2.0.0 release before the mainnet release? If so can you reach out to Marcelo?

**Comment by staffik** [comment]
> What about validator key hot swap? Should this check also be triggered then?
It would be nice to have this check as well, will add it.

> We also discussed forcing single shard tracking for stateless validators - are you going to do that separately?
Yes, it will be done separately.
 
> Should this be cherry picked into the 2.0.0 release before the mainnet release? If so can you reach out to Marcelo?
Yes, will reach out to Marcelo after this PR is merged.

File: chain/client/src/client_actor.rs [hunk 3]
```diff
@@ -1170,6 +1188,9 @@ impl ClientActorInner {
                 &|validator_signer| self.client.update_validator_signer(validator_signer),
             );
             if update_result.validator_signer_updated {
+                if let Err(error) = check_validator_tracked_shards(&self.client) {
+                    error!(target: "client", ?error, "Could not check validator tracked shards");
+                }
```
Comment (wacban) [inline chain/client/src/client_actor.rs:1193]: I'd be tempted to turn that into an `expect`. WDYT?

**Review by wacban — APPROVED** [review]
LGTM

**Review by tayfunelmas — APPROVED** [review]

**Pull request merged**